Feature ImageData in SpeedyMediaSource

kalwalt opened the pull request:
Continue from PR #56

kalwalt committed:
New class SpeedyImageDataMediaSource to handle ImageData
- added new ImageData MediaType

kalwalt commented on types/core/speedy-media-source.d.ts:
```diff
@@ -30,7 +30,7 @@ export class SpeedyMediaSource {
      * The underlying wrapped object
      * @returns {SpeedyMediaSourceNativeElement}
      */
-    get data(): SpeedyMediaSourceNativeElement;
```
I don't know why the typescript compiler convert to any, instead it should be SpeedyMediaSourceNativeElement

kalwalt commented on types/core/speedy-media-source.d.ts:
```diff
@@ -73,5 +73,5 @@ export class SpeedyMediaSource {
     _load(element: SpeedyMediaSourceNativeElement): SpeedyPromise<SpeedyMediaSource>;
 }
-export type SpeedyMediaSourceNativeElement = HTMLImageElement | HTMLVideoElement | HTMLCanvasElement | ImageBitmap;
+export type SpeedyMediaSourceNativeElement = HTMLImageElement | HTMLVideoElement | HTMLCanvasElement | any | ImageBitmap | ImageData;
```
Tsc export OffscreenCanvas as any instead ImageData is correct. Maybe OffscreenCanvas should be added to MediaType?

alemart commented:
Nice! We can just change files src/core/speedy-media-source.js and src/utils/types.js. Let's leave the other files, including the typescript definitions, for a different commit. We haven't changed the image sink yet.

alemart commented on src/core/speedy-media-source.js:
```diff
@@ -710,3 +712,102 @@ class SpeedyBitmapMediaSource extends SpeedyMediaSource
+     */
+    constructor(token) {
+        super(token);
+        this._data = null;
```
We can annotate the type here for consistency, as in `/** @type {ImageData} image data */`

alemart commented on src/core/speedy-media-source.js:
```diff
@@ -710,3 +712,102 @@ class SpeedyBitmapMediaSource extends SpeedyMediaSource
+     * @returns {null}
+     */
+    release() {
+        if (this._data != null)
+            this._data.close();
+
```
Unlike an `ImageBitmap`, an `ImageData` does not have a `close()` method. We should just remove these lines.

kalwalt replied:
Ok i will remove It.

kalwalt replied:
So it should be:

```js
 release() {
        return super.release();
    }
```
does it correct?

alemart replied:
It is correct, but since this method does nothing, we can just remove it altogether. As you can see, most other media sources don't implement it.

kalwalt replied:
Ouch! You are right i will  when  i came back from holiday.🙂

alemart commented:
In addition, in `SpeedyMedia.type`, let's handle the case in which the image source is an `ImageData`. Let's make it return the string `'data'`:

https://github.com/alemart/speedy-vision/blob/48934ac8317a5b7c3a665af00c69d22e626b888a/src/core/speedy-media.js#L113-L138

kalwalt committed:
fix for reviews from @alemart
- added @type to _data in the SpeedyImageDataMediaSource constructor
- return string 'data' for ImageData MediaType in the get() method

alemart commented:
Hi @kalwalt,

Please sign your commits - see [CONTRIBUTING.md](https://github.com/alemart/speedy-vision/blob/master/CONTRIBUTING.md).

I ask you to remove all files from this PR, except:
* src/core/speedy-media-source.js
* src/core/speedy-media.js
* src/utils/types.js

kalwalt commented:
> Hi @kalwalt,
> 
> Please sign your commits - see [CONTRIBUTING.md](https://github.com/alemart/speedy-vision/blob/master/CONTRIBUTING.md).
> 
> I ask you to remove all files from this PR, except:
> 
> * src/core/speedy-media-source.js
> * src/core/speedy-media.js
> * src/utils/types.js

Maybe at this point it's better to recreate it. I'm using a GPG key to sign my commits but i will do as you mentioned in the contributing page. Sorry for this!

kalwalt commented:
New updated PR #58

alemart commented:
I think that it will be a better design to modify the existing image sink to introduce support for exporting an `ImageData` instead of creating a new node. I'll be doing that later.
